### PR TITLE
DAP-8 Login Functionality

### PR DESCRIPTION
**Pull request by Marcelkibor**
# Description of What I changed

Issue I worked on

see https://ampath.atlassian.net/browse/DAP-2

# Checklist:

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my code
- [x] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes
- [ ] Any dependent changes have been merged and published in downstream modules

### CODE REVIEW

File: .gitignore [hunk 1]
```diff
@@ -22,3 +22,5 @@ dist-ssr
 *.njsproj
 *.sln
 *.sw?
+.yarn
+.env
```
Comment (kipchubett) [inline .gitignore]: https://gist.github.com/OleksiyRudenko/d51388345ea55767b7672307fe35adf3

File: src/authentication/Login/LoginRequest.tsx [hunk 2]
```diff
@@ -0,0 +1,39 @@
...
+        redirect: 'follow',
+        });
+        const [headers, body] = await Promise.all([
+        result.headers,
+        result.json(),
+        ]);
+        if (body.authenticated) {
+            storage.saveInfo(body);
+            localStorage.setItem('authenticated', body.authenticated);
+            localStorage.setItem('Banner', 'true');
```
Comment (kipchubett) [inline src/authentication/Login/LoginRequest.tsx]: What is the use for the item `Banner`?

File: src/authentication/Login/LoginRequest.tsx [hunk 3]
```diff
@@ -0,0 +1,39 @@
...
+            window.location.href = "/"
+        } 
+        else {
+            response.error = 'Invalid username or password';
+        }
+} 
+    catch (e) {
+        response.error = 'Network error';
```
Comment (kipchubett) [inline src/authentication/Login/LoginRequest.tsx]: This is a catch-all error. The above request can fail due to different reasons e.g 504 gateway timeout or server error with status code 500.  Please return the actual error and handle different status codes

File: src/authentication/Login/Login.tsx [hunk 4]
```diff
@@ -0,0 +1,59 @@
...
+import storage from "../../app/localStorage";
+import Logo from '../../public/am-logo.jpg';
+import React from "react";
+import ClipLoader from "react-spinners/ClipLoader";
+import LoginRequest from "./LoginRequest";
+const Login = () => {
+const userInformation = localStorage.getItem("authenticated")
+if(userInformation == "true"){
+    window.location.href = "/"
+}
```
Comment (kipchubett) [inline src/authentication/Login/Login.tsx]:
```suggestion
const isAuthenticated = localStorage.getItem("authenticated")
if(isAuthenticated && isAuthenticated === "true"){
    window.location.href = "/"
}
```

File: src/mocks/handlers.ts [hunk 5]
```diff
@@ -0,0 +1,32 @@
+import { rest, RestRequest } from 'msw';
+
+export const handlers = [
+  rest.get('https://dev3.openmrs.org/openmrs/ws/rest/v1/session', (req, res, ctx) => {
```
Comment (kipchubett) [inline src/mocks/handlers.ts]: You shouldn't be hard-coding base URLs. Extract this to constant or a test URL e.g localhost

File: src/authentication/Login/LoginRequest.tsx [hunk 6]
```diff
@@ -0,0 +1,39 @@
...
+        headers: {
+            Authorization: 'Basic ' + btoa(username + ':' + password),
+        },
+        method: 'GET',
+        redirect: 'follow',
+        });
+        const [headers, body] = await Promise.all([
+        result.headers,
+        result.json(),
+        ]);
```
Comment (kipchubett) [inline src/authentication/Login/LoginRequest.tsx]: I would recommend you extract this into an `authentication.resource.ts` file

**Review by kipchubett — CHANGES_REQUESTED** [review]
Good progress @Marcelkibor 

I've made some suggestions and recommendations please go thru them and make the necessary changes

File: src/authentication/Login/LoginRequest.tsx (hunk 3 above)
Comment (Marcelkibor) [reply src/authentication/Login/LoginRequest.tsx]: Resolved.

File: src/authentication/Login/LoginRequest.tsx (hunk 6 above)
Comment (Marcelkibor) [reply src/authentication/Login/LoginRequest.tsx]: Resolved

File: src/authentication/Login/Login.tsx (hunk 4 above)
Comment (Marcelkibor) [reply src/authentication/Login/Login.tsx]: Resolved. Transferred the logic to protected-routes component

File: src/mocks/handlers.ts (hunk 5 above)
Comment (Marcelkibor) [reply src/mocks/handlers.ts]: Resolved

**Review by kipchubett — APPROVED** [review]

**Pull request merged**